### PR TITLE
pet-charity-be-feature/registration - created user route/schema

### PR DESCRIPTION
**Pull request by alexander-svirzhevsky**
Realized a registration

### CODE REVIEW

File: config/default.json [hunk 1]
```diff
@@ -1,3 +1,4 @@
 {
-    "mongoURI": "mongodb+srv://alexander123:alexander123@cluster.7nzur.mongodb.net/myFirstDatabase?retryWrites=true&w=majority"
+    "mongoURI": "mongodb+srv://alexander123:alexander123@cluster.7nzur.mongodb.net/myFirstDatabase?retryWrites=true&w=majority",
```
Comment (universezzz) [inline config/default.json:2]: Why is your secret config public? :DDD

File: routes/api/users.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+const { check, validationResult } = require("express-validator");
+const config = require("config");
+const gravatar = require("gravatar");
+const argon2 = require("argon2");
+const jwt = require("jsonwebtoken");
+const User = require("../../db/Schema/User");
+
+// @route     POST api/users
+// @desc      Register user
+// @access    Publick
```
Comment (universezzz) [inline routes/api/users.js:12]: Publick :DD
Remove all those comments, you'll add swagger soon, you don't need this anymore.

P.S.
// @route     POST api/users
// @desc      Register user
// @access    Publick 
remove this everywhere

File: routes/api/users.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
+const express = require("express");
+const router = express.Router();
+const { check, validationResult } = require("express-validator");
+const config = require("config");
+const gravatar = require("gravatar");
+const argon2 = require("argon2");
+const jwt = require("jsonwebtoken");
+const User = require("../../db/Schema/User");
```
Comment (universezzz) [inline routes/api/users.js:8]: always separate libraries and your imports. It makes your code clean and more readable
e.g. 
`
const express = require("express");
const router = express.Router();
const { check, validationResult } = require("express-validator");
const config = require("config");
const gravatar = require("gravatar");
const argon2 = require("argon2");
const jwt = require("jsonwebtoken");

const User = require("../../db/Schema/User");
`

File: routes/api/users.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
+const express = require("express");
+const router = express.Router();
+const { check, validationResult } = require("express-validator");
+const config = require("config");
+const gravatar = require("gravatar");
```
Comment (universezzz) [inline routes/api/users.js:5]: you don't need gravatar at all, remove this package (npm unistall gravatar). It's a reason why I told you create a separate branch for FE/BE initial structure.

File: routes/api/users.js [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+  async (req, res) => {
+    const errors = validationResult(req);
+    if (!errors.isEmpty()) {
+      return res.status(400).json({ errors: errors.array() });
+    }
+
+    const { name, email, password } = req.body;
+
+    try {
+      // See if user exists
```
Comment (universezzz) [inline routes/api/users.js:32]: get rid of it

File: routes/api/users.js [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+          .status(400)
+          .json({ errors: [{ msg: "User already exists" }] });
+      }
+
+      // Get users gravatar
+      const avatar = gravatar.url(email, {
+        s: "200",
+        r: "pg",
+        d: "mm",
+      });
```
Comment (universezzz) [inline routes/api/users.js:46]: get rid of it

File: db/Schema/User.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    required: true,
+    unique: true,
+  },
+  password: {
+    type: String,
+    required: true,
+  },
+  avatar: {
+    type: String,
+  },
```
Comment (universezzz) [inline db/Schema/User.js:19]: get rid of it

File: routes/api/users.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+      user = new User({
+        name,
+        email,
+        password,
+        avatar,
+      });
+
+      // Encrypt password Argon2
```
Comment (universezzz) [inline routes/api/users.js:55]: get rid of it

File: routes/api/users.js [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+      await argon2.hash(password);
```
Comment (universezzz) [inline routes/api/users.js:56]: wtf is this? You must store **hashed** password in database. But you create an user on lane 48, and after that just hash password, you don't even get result from argon2.hash function. 

`
const hashedPassword = await argon2.hash(password);

user = new User({
  name,
  email,
  password: hashedPassword,
});
      
await user.save();
`

File: routes/api/users.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+        { expiresIn: 360000 },
+        (err, token) => {
+          if (err) throw err;
+          res.json({ token });
+        }
+      );
+    } catch (error) {
+      console.error(error.message);
+      res.status(500).send("Server error");
+    }
```
Comment (universezzz) [inline routes/api/users.js:79]: You have to create a **The default error handler**
https://expressjs.com/en/guide/error-handling.html

File: server.js [hunk 11]
```diff
@@ -3,12 +3,17 @@ const app = express();
 const PORT = 5000;
 const connectDB = require("./db/db");
 
+// Connect Database
```
Comment (universezzz) [inline server.js:6]: get rid of it

File: server.js [hunk 12]
```diff
@@ -3,12 +3,17 @@ const app = express();
 const PORT = 5000;
```
Comment (universezzz) [inline server.js:3]: const PORT = process.env.PORT || 5000;
In production build you'll specify port based on environment variable.

File: server.js [hunk 13]
```diff
@@ -3,12 +3,17 @@ const app = express();
...
 const connectDB = require("./db/db");
 
+// Connect Database
 connectDB();
 
+app.use(express.json({ extended: false }));
+
 app.get("/", (req, res) => {
   res.send("Hello!");
 });
```
Comment (universezzz) [inline server.js:13]: get rid of it

**Review by universezzz — CHANGES_REQUESTED** [review]

File: config/default.json (hunk 1 above)
Comment (alexander-svirzhevsky) [reply config/default.json:2]: where should he be? :)

Comment (alexander-svirzhevsky) [reply config/default.json:2]: add config folder in gitignore

File: routes/api/users.js (hunk 2 above)
Comment (alexander-svirzhevsky) [reply routes/api/users.js:12]: remove all comments

File: routes/api/users.js (hunk 3 above)
Comment (alexander-svirzhevsky) [reply routes/api/users.js:8]: separate libraries and my imports

File: routes/api/users.js (hunk 4 above)
Comment (alexander-svirzhevsky) [reply routes/api/users.js:5]: uninstall gravatar

File: routes/api/users.js (hunk 5 above)
Comment (alexander-svirzhevsky) [reply routes/api/users.js:32]: remove all comments

File: routes/api/users.js (hunk 9 above)
Comment (alexander-svirzhevsky) [reply routes/api/users.js:56]: got it, my mistake ....

File: server.js (hunk 11 above)
Comment (alexander-svirzhevsky) [reply server.js:6]: got it

File: server.js (hunk 12 above)
Comment (alexander-svirzhevsky) [reply server.js:3]: got it, and fixed it

File: server.js (hunk 13 above)
Comment (alexander-svirzhevsky) [reply server.js:13]: ok

File: routes/api/users.js (hunk 10 above)
Comment (alexander-svirzhevsky) [reply routes/api/users.js:79]: "is added at the end of the middleware function stack" 
using this i fixed problem, but not shure if it's right
users js - pet-charity-be - Visual Studio Code  Ad

**Pull request merged**